### DISCUSSION TITLE
Google Adsense  Question

### DISCUSSION THREAD

**Discussion by JamesSingleton** [vercel/next.js#32323] [Help]
So I have looked through the one or two questions already here about Google AdSense but nothing seems "definitive" in my opinion. Some same you should add this script to the `_document.tsx`

```tsx
import Document, { Head, Html, Main, NextScript } from 'next/document'

class MyDocument extends Document {
  render() {
    const ADSENSE_ID = process.env.NEXT_PUBLIC_ADSENSE_ID
    return (
      <Html lang="en-US" className="h-full bg-gray-100">
        <Head>
          <script
            async
            src={`https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?${ADSENSE_ID}`}
          />
        </Head>
        <body className="loading h-full">
          <Main />
          <NextScript />
        </body>
      </Html>
    )
  }
}

export default MyDocument
```

And then put the `.push({})` in the Ad Component you make like:
```tsx
import { useEffect, FC } from 'react'

declare global {
  interface Window {
    adsbygoogle: { [key: string]: unknown }[]
  }
}

interface AdBannerProps {
  adSlot: number
}

const AdBanner: FC<AdBannerProps> = ({ adSlot }) => {
  const ADSENSE_ID = process.env.NEXT_PUBLIC_ADSENSE_ID
  useEffect(() => {
    if (typeof window !== 'undefined') {
      ;(window.adsbygoogle = window.adsbygoogle || []).push({})
    }
  }, [])
  return (
    <ins
      className="adsbygoogle adbanner-customize"
      style={{
        display: 'block',
      }}
      data-ad-client={ADSENSE_ID}
      data-ad-slot={adSlot}
      data-ad-format="auto"
      data-full-width-responsive="true"
    />
  )
}

export default AdBanner
```

But then there is the suggestion of using `next/script` and adding both of those to the `_app.tsx` file:
```tsx
import Script from 'next/script'

const App = ({ Component, pageProps }) => {
  const ADSENSE_ID = process.env.NEXT_PUBLIC_ADSENSE_ID

  return (
    <>
      {/* Global Site Tag (gtag.js) - Google Analytics */}
      <Script
        strategy="afterInteractive"
        src={`https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?${ADSENSE_ID}`}
      />
      <Script
        id="adsbygoogle-init"
        strategy="afterInteractive"
        dangerouslySetInnerHTML={{
          __html: `
          (adsbygoogle = window.adsbygoogle || []).push({
            google_ad_client: {ADSENSE_ID},
            enable_page_level_ads: true
       });
          `,
        }}
      />
      <Component {...pageProps} />
    </>
  )
}

export default App
```

Google says that `pagead2` script needs to be in the `<head />` so my assumption is to add it in the `_document.tsx`... But I am unsure about the other part.

**Comment by ghost** [comment]
same issue here getting adsbygoogle.js?client=ca-pub-XXXXXXXXXXXXXXXX AdSense head tag doesn't support data-nscript attribute.

**Reply by permlap-phola** [reply]
Can you find the solution for this?

**Reply by MarianoJimenezPerez** [reply]
did you resolve it?

**Comment by AibSyed** [comment]
@JamesSingleton was this issue ever resolved? I'm currently having the same problem.

**Reply by JamesSingleton** [reply]
I never got it working for me, for whatever reason Google Ads doesn't want to put ads on my website that averages about 7,700 page views a month 🤷🏻‍♂️ so I removed the code from my application.

**Comment by Focus-me34** [comment]
🆙 
I'm having the same issue as you guys did 3 years ago... "AdSense head tag doesn't support data-nscript attribute." is the error I get in the logs of my Next app (app router).
I'm using the Script component provided by nextJs. I'm surprised that this issue hasn't been solved yet. I just started investigating though. I'll come back with the answer if I find any 👍🏼

**Reply by AibSyed** [reply]
> 🆙 I'm having the same issue as you guys did 3 years ago... "AdSense head tag doesn't support data-nscript attribute." is the error I get in the logs of my Next app (app router). I'm using the Script component provided by nextJs. I'm surprised that this issue hasn't been solved yet. I just started investigating though. I'll come back with the answer if I find any 👍🏼

Not sure what version of Next.js you're currently on, but when I was dealing with this, I was running v13.4.7 with the Pages Router.

What worked for me was adding the AdSense `<script>` directly in `_document.js` like this:

```
import Document, { Html, Head, Main, NextScript } from 'next/document';

export default class MyDocument extends Document {
  render() {
    return (
      <Html lang="en">
        <Head>
          <meta charSet="UTF-8" />
          {/* Adding Google AdSense script directly */}
          <script
            src={`https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=${process.env.NEXT_PUBLIC_GOOGLE_ADS_CLIENT_ID}`}
            async
          />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```

By adding the AdSense script directly in the `<Head>`, I bypassed the Next.js `Script` component, which usually adds a `data-nscript` attribute that AdSense doesn’t support. This approach worked without errors.

Hope this helps!

**Reply by caio303** [reply]
@AibSyed had the same problem here and your fix worked for me, thanks!

**Reply by AibSyed** [reply]
> @AibSyed had the same problem here and your fix worked for me, thanks!

Awesome, happy to help!
